gymnasium/utils/serialize_spec_stack.py

```diff
@@ -0,0 +1,139 @@
+"""Functions that use/modify specification stacks."""
+import dataclasses
+import inspect
+import json
+import re
+from typing import Union
+
+from gymnasium import error
+from gymnasium.envs.registration import EnvSpec
+from gymnasium.wrapperspec import WrapperSpec
+
+
+def serialise_spec_stack(stack: "tuple[Union[WrapperSpec, EnvSpec]]") -> str:
+    """Serialises the specification stack into a JSON string.
+
+    Args:
+        stack: Tuple of environment and wrapper specifications, known as the specification stack. Generated via `env.spec_stack`.
+
+    Returns:
+        A JSON string representing the specification stack.
+    """
+    num_layers = len(stack)
+    stack_json = {}
+    for i, spec in enumerate(stack):
+        spec = _serialise_callable(spec)
+        if i == num_layers - 1:
+            if isinstance(spec, WrapperSpec) or not isinstance(
+                spec, EnvSpec
+            ):  # NB: WrapperSpec is a subclass of EnvSpec
+                raise error.Error(
+                    f"Expected spec to be an EnvSpec, but got {type(spec)} instead."
+                )
+            layer = "raw_env"
+        else:
+            if not isinstance(spec, WrapperSpec):
+                raise error.Error(
+                    f"Expected spec to be a WrapperSpec, but got {type(spec)} instead."
+                )
+            layer = f"wrapper_{num_layers - i - 2}"
+        spec_json = json.dumps(dataclasses.asdict(spec))
+        stack_json[layer] = spec_json
+    return stack_json
+
+
+def _serialise_callable(spec: Union[WrapperSpec, EnvSpec]) -> dict:
```

Comment:
(forgot to finish the comment in the review)

As I understand, the goal here is to serialize a (more or less arbitrary) python function into a string? 

This is, in general, borderline impossible in a sane way. There are some approaches and some libraries which do that, but there are generally some limitations. Importantly, I don't think we can hope to find a way to do it robustly, so I'd strongly suggest reconsidering the design that requires this functionality. (unless we specify the exact limitations and support everything within those bounds) If we do try to do it, we almost definitely don't want to try and make our own implementation 

To eliminate a few tempting options:
- The current implementation doesn't work because it sometimes parses incorrectly, and it only could work with `lambda x: y` functions, which is a very bad limitation (it also has other important limitations)
- Getting the full function code doesn't work because it might use externally defined functions
- Saving and executing the full file doesn't work because it might have side effects, also there could be external imports
- Saving and executing the full file and all files that it imports doesn't work because that might include huge external libraries (e.g. numpy)
- Saving and executing the full file and all files that it imports, checking if the other module is defined by the user, and otherwise just copying the import so that the appropriate library is imported after deserialization is actually Einsten's definition of insanity (the whole repetition thing is a secondary symptom).

Source: I tried to do the same thing some years ago. Don't recommend.